### PR TITLE
V1.1.dev dlt

### PR DESCRIPTION
**Pull request by debthomas**
Updated for comments of moving the code to the timesheet library.

### CODE REVIEW

File: Timesheet.gs [hunk 1]
```diff
@@ -39,18 +39,25 @@ var Log_
...
-
-  checkIn:                 [function() {},  'checkIn()',                'checkIn failed',                       checkIn_],
-  checkOut:                [function() {},  'checkOut()',               'checkOut failed',                      checkOut_],
+//                         Name                         onError Message                        Main Functionality
+//                         ----                         ---------------                        ------------------
+  onInstall:               ['onInstall()',              'Failed to install',                    onInstall_],
+  initialize:              ['initialize()',             'Failed to initialize',                 initialize_],
+  checkIn:                 ['checkIn()',                'checkIn failed',                       checkIn_],
+  checkOut:                ['checkOut()',               'checkOut failed',                      checkOut_],
+  isCheckedIn:             ['isCheckedIn()',            'isCheckedIn failed',                   isCheckedIn_],
```
Comment (andrewroberts) [inline Timesheet.gs:47]: For a moment I wondered where getStatus() had gone :)

File: Timesheet.gs [hunk 2]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
-    if (PRODUCTION_VERSION_) {
+    Assert.handleError(assertConfig) 
+  }
+  
+} // eventHandler_()
+
+// Private event handlers
+// ----------------------
+
+function onInstall_(event){
```
Comment (andrewroberts) [inline Timesheet.gs]: You can delete this function. It is only used for Add-ons.

File: Timesheet.gs [hunk 3]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
-      });
+function onOpen_(event) {
     
+  var menu = SpreadsheetApp
+        .getUi().createMenu('[ Timesheet ]')
+        
+      menu.addItem('Check In',  'checkIn')
+      menu.addItem('Check Out', 'checkOut')
+        
+  if (event.authMode === ScriptApp.AuthMode.NONE) {
```
Comment (andrewroberts) [inline Timesheet.gs]: This can be deleted as it only happens in an Add-on.

File: Timesheet.gs [hunk 4]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
-      Log_ = BBLog.getLog({
-        displayUserId:        BBLog.DisplayUserId.USER_KEY_FULL,
-        lock:                 lock,
-        firebaseUrl:          firebaseUrl,
-        firebaseSecret:       firebaseSecret,
-      });
+function onOpen_(event) {
     
+  var menu = SpreadsheetApp
+        .getUi().createMenu('[ Timesheet ]')
```
Comment (andrewroberts) [inline Timesheet.gs:145]:
```suggestion
    .getUi().createMenu('[ Timesheet ]')
```

File: Timesheet.gs [hunk 5]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
+        
+      menu.addItem('Check In',  'checkIn')
+      menu.addItem('Check Out', 'checkOut')
```
Comment (andrewroberts) [inline Timesheet.gs]:
```suggestion
  menu.addItem('Check Out', 'checkOut')
```

File: Timesheet.gs [hunk 6]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
-        lock:                 lock,
-        firebaseUrl:          firebaseUrl,
-        firebaseSecret:       firebaseSecret,
-      });
+function onOpen_(event) {
     
+  var menu = SpreadsheetApp
+        .getUi().createMenu('[ Timesheet ]')
+        
+      menu.addItem('Check In',  'checkIn')
```
Comment (andrewroberts) [inline Timesheet.gs]:
```suggestion
  menu.addItem('Check In',  'checkIn')
```

File: Timesheet.gs [hunk 7]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
+      menu.addItem('Check Out', 'checkOut')
+        
+  if (event.authMode === ScriptApp.AuthMode.NONE) {
+   
+    menu.addItem('Start', 'initialize') 
+
+  } else { // LIMITED or FULL
+  
+    var triggerOpenId = PropertiesService.getDocumentProperties().getProperty('triggerOpenId')
```
Comment (andrewroberts) [inline Timesheet.gs]: This should be using the doc props from the bound script, like where the "checked in" field is stored.
```suggestion
    var triggerOpenId = properties.getProperty('triggerOpenId')
```

File: Timesheet.gs (hunk 5 above)
Comment (andrewroberts) [reply Timesheet.gs]: ... and don't show these until the "Start" has been processed.

File: Timesheet.gs [hunk 8]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
+        
+    if (triggerOpenId === null) {
+  
+      menu.addItem('Start', 'initialize')   
+  
     } else {
+    
+      var html = HtmlService
+      .createHtmlOutputFromFile("index")
```
Comment (andrewroberts) [inline Timesheet.gs:159]:
```suggestion
        .createHtmlOutputFromFile("index")
```

File: Timesheet.gs [hunk 9]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
+      .setTitle("Check In/Check Out");
```
Comment (andrewroberts) [inline Timesheet.gs:160]:
```suggestion
        .setTitle("Check In/Check Out");

```

File: Timesheet.gs [hunk 10]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
+      SpreadsheetApp.getUi().showSidebar(html)
+    }    
+  }
+  menu.addToUi()
```
Comment (andrewroberts) [inline Timesheet.gs:164]:
```suggestion

  menu.addToUi()
```

File: Timesheet.gs [hunk 11]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
-  } // eventHandler_.initialseEventHandler() 
+  var status = properties.getProperty("TIMESHEET_STATUS")
+  Log_.fine('status: ' + status)
 
-} // eventHandler_()
+  if (status === "CHECKED_IN"){
+    return true
+  } else {
+    return false
+  }
```
Comment (andrewroberts) [inline Timesheet.gs]: tip: This could be 

    return (status === 'CHECKED_IN' ? true : false

File: Timesheet.gs [hunk 12]
```diff
@@ -67,134 +74,152 @@ function checkOut (arg1, arg2, properties, lock) {return eventHandler_(EVENT_HAN
...
+}
 
-// Private event handlers
-// ----------------------
+function initialize_() {
+
+  var properties = PropertiesService.getDocumentProperties()
```
Comment (andrewroberts) [inline Timesheet.gs]: Again using library rather than bound script doc props

File: index.html [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    <script>
+    
+    // Initially Set the button status
+    setButtonStatus();
+    
+    /* Run the setButtonStatus to set the text on the button */              
+    function setButtonStatus() {
+    
+      console.log("In setButtonStatus")
+      var checkedIn = google.script.run.withSuccessHandler(setText).withFailureHandler(function(err){ alert("Error on isCheckedIn")}).isCheckedIn();
```
Comment (andrewroberts) [inline index.html]: That works but is kinda clunky and you are not displaying the error. 

Add a hidden div with id="error" then set the value to the error message.
   
    <input ...
    <div id="error" hidden></div>

    function errorHandler(error) { // Same one for all the "script.google.run"s
      document.getElementById("error").value = error.message
      document.getElementById("error").style.display = "inline" // "none" to hide it again
    }

File: index.html [hunk 14]
```diff
@@ -0,0 +1,89 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <base target="_top">
+    <link rel="stylesheet" href="https://ssl.gstatic.com/docs/script/css/add-ons1.css">
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
```
Comment (andrewroberts) [inline index.html]: You can delete this line as you aren't using any jQuery.

File: index.html [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        
+      } // setButtonStatus.setText(checkedIn)
+      
+    } // setButtonStatus()
+     
+    function checkingInOut() {
+
+      console.log("In checkingInOut")      
+      var checkedIn = google.script.run.withSuccessHandler(runCheckInOut).withFailureHandler(function(err){ alert("Error on isCheckedIn")}).isCheckedIn();
```
Comment (andrewroberts) [inline index.html]:
```suggestion
      var checkedIn = google.script.run
        .withSuccessHandler(runCheckInOut)
        .withFailureHandler(function(err) {
          alert("Error on isCheckedIn")
        })
        .isCheckedIn();
```

File: index.html [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+      function runCheckInOut(checkedIn) {
+ 
+        console.log("In runCheckInOut: " + checkedIn)      
+ 
+        if (!checkedIn && document.getElementById("checkinout").value === "Check In") {
+        
+          /* Runs the serverside script checkIn 
+          * Then calls the "setButtonStatus" function.
+          */
+          google.script.run.withSuccessHandler(setButtonStatus).withFailureHandler(function(err){ alert("Error on checkIn")}).checkIn();
```
Comment (andrewroberts) [inline index.html]: Indent as above.

File: index.html [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        
+        } else if (checkedIn && document.getElementById("checkinout").value === "Check Out"){
+        
+          /* Runs the serverside script checkIn 
+          * Then calls the "setButtonStatus" function.
+          */
+          google.script.run.withSuccessHandler(setButtonStatus).withFailureHandler(function(err){ alert("Error on checkOut")}).checkOut();
```
Comment (andrewroberts) [inline index.html]: Indent as above

File: index.html [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+     <br>
+     <input class="action" id="checkinout" type="button" value="" onmouseup="checkingInOut()">
+    </div>
+     
+    <script>
+    
+    // Initially Set the button status
+    setButtonStatus();
+    
+    /* Run the setButtonStatus to set the text on the button */
```
Comment (andrewroberts) [inline index.html]:
```suggestion
    // Run the setButtonStatus to set the text on the button           
```

File: index.html [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        console.log("In runCheckInOut: " + checkedIn)      
+ 
+        if (!checkedIn && document.getElementById("checkinout").value === "Check In") {
+        
+          /* Runs the serverside script checkIn 
+          * Then calls the "setButtonStatus" function.
+          */
+          google.script.run.withSuccessHandler(setButtonStatus).withFailureHandler(function(err){ alert("Error on checkIn")}).checkIn();
+        
+        } else if (checkedIn && document.getElementById("checkinout").value === "Check Out"){
```
Comment (andrewroberts) [inline index.html]:
```suggestion
        } else if (checkedIn && document.getElementById("checkinout").value === "Check Out") {
```

**Review by andrewroberts — CHANGES_REQUESTED** [review]
Looking good. Is mostly working on this test sheet: https://docs.google.com/spreadsheets/d/1GfkBGfzxfn7PXxUA9JLjhq1qaey7dTIbuxpyF37SFCo/edit#gid=0.

The "check in" from the menu and the sidebar don't sync. "Check in" should change to "Check out" on button if user checks in from menu.

And it would be nice if there was also only one menu option which automatically updated depending on the "checked in" status. I often click "check in" when I've forgotten to check out.

Haven't checked it on the template you were talking about - can you send me a link.

File: Timesheet.gs (hunk 1 above)
Comment (debthomas) [reply Timesheet.gs:47]: ;-)

File: Timesheet.gs (hunk 2 above)
Comment (debthomas) [reply Timesheet.gs]: Ah, I didnt think this was used, but left in for completeness of the template, will remove.

File: Timesheet.gs (hunk 3 above)
Comment (debthomas) [reply Timesheet.gs]: Ok I have removed the  if (event.authMode === ScriptApp.AuthMode.NONE) { and assosiated code. 

The staff folder had it, does staff data have add-ons?

File: Timesheet.gs (hunk 5 above)
Comment (debthomas) [reply Timesheet.gs]: Moved the Check In to the else where the sidebar is displayed

File: Timesheet.gs (hunk 6 above)
Comment (debthomas) [reply Timesheet.gs]: done

File: Timesheet.gs (hunk 7 above)
Comment (debthomas) [reply Timesheet.gs]: Ah yes.

File: Timesheet.gs (hunk 10 above)
Comment (debthomas) [reply Timesheet.gs:164]: Done

File: Timesheet.gs (hunk 11 above)
Comment (debthomas) [reply Timesheet.gs]: Changed

File: Timesheet.gs (hunk 12 above)
Comment (debthomas) [reply Timesheet.gs]: Changed

File: index.html (hunk 13 above)
Comment (debthomas) [reply index.html]: Yes that is clearer.

File: index.html (hunk 14 above)
Comment (debthomas) [reply index.html]: Removed, I got this from a example.

File: index.html (hunk 16 above)
Comment (debthomas) [reply index.html]: Done

File: index.html (hunk 17 above)
Comment (debthomas) [reply index.html]: done

File: index.html (hunk 18 above)
Comment (debthomas) [reply index.html]: Done

File: index.html (hunk 19 above)
Comment (debthomas) [reply index.html]: Done

File: index.html (hunk 15 above)
Comment (debthomas) [reply index.html]: Done

**Comment by debthomas** [comment]
Applied comments except the syncing of sidebar with menu and changing the menu to be one menu option instead of 2.

File: Timesheet.gs (hunk 3 above)
Comment (andrewroberts) [reply Timesheet.gs]: There's a story! 

tl;dr. It was an Add-on, but we've reverted to bound script for now. 

Chad was hoping to sell the software he'd developed whilst working at CCN to other churches, then a new boss was due to start in Jan 2019 and Chad was terrified he'd pull the plug on CloudFire (as he called it). So there was a a mad flurry of activity at the end of last year as I converted all the components into Add-ons so other churches could use them. Fast forward to now: Chad never managed to sell CloudFire, and the new boss left after a few months anyway. Updating Add-ons is harder than doing it in a bound script so we've reverted to bound scripts for now, but are leaving the Add-on bits (onInstall(), etc) for when he does finally start selling CloudFire.
